### ISSUE TITLE
Slow Unittest Startup

### ISSUE THREAD

**Issue by nbartels** [gwt-test-utils/gwt-test-utils#81]
After the migration from GWT 2.5 to 2.7 and the corresponding migration to the newest `gwt-test-utils` version the unit test are really slow. Before the migration they ran for some seconds (< 10s) and now *no* test is running under 20 secs.

I know that there are major changes because of the overlay types but nevertheless I tried to find the performance sink.

So I added some log output to the `GwtFactory` and found some interesting thing and hopefully someone can explain or better fix this :smile: 

Every step in the GwtFactory init process is measured and the slow part is coming from a inner GWT class that is called `CompileMoreLater`.

Here is a overview, if need I can add some explanation. 

```
Building temp CompilerContext took ms: 2
Building moduleDef took ms: 570
Building CompilerContext took ms: 1734
CompilationState - Post TreeLogger 0
 + Compiling resources : 3521
 + Compiling took ms:  16
 +- jar:file:/D:/sources/XXX/lib/validation-api-1.0.0.GA-sources.jar found times: 45
 +- jar:file:/D:/sources/XXX/lib/gwt-dev-2.7.0.jar found times: 22
 +- jar:file:/D:/sources/XXX/war/WEB-INF/lib/gxt-2.3.1-gwt22.jar found times: 502
 +- jar:file:/D:/sources/XXX/lib/gwt-user-2.7.0.jar found times: 2913
 +- jar:file:/D:/sources/XXX/war/WEB-INF/lib/gwt-log-3.2.2-customized.jar found times: 39
 + Start with 'compile more later'
 ++ CompilationUnitBuilders processed: 3521
 + 'Compile more Later' finished, took ms:  13768
CompilationState - Post state 14763
Building CompilationState took ms: 14764
Building overlayRewriter took ms: 16
Building gwtClassLoader took ms: 7
GWTFactory init 17519
```

As you can see the jars are recompiled and the complete init needs about 17 secs. I checked some tests and you get the same picture per test class.

GXT is a customized version so it runs with gwt 2.7, so this is no problem.

I don't understand why the created artifacts are not stored somewhere and reused on a second test run.

Any suggestions?

**Comment by nbartels** [comment]
So, I checked it a bit more because the issue is really annoying and the problem is the `UnitCache`. ATM the `MemoryUnitCache` is used this is rebuild with every UnitTest class. Depending on the use additional gwt libs (like gwt-log) this may take some time.

My very custom solution is a custom unit cache 😄 

It saves a list of `CompilationUnit` items on cleanup as serialised `ArrayList` to the hard disc and in the constructor the `ArrayList` is read from disc again. So the first test class is a bit slow and the following are fast again. 

To solve the problem of outdated compilation units, a 15 minute threshold is added and after that period of time the normal compilation process takes place again. Between two test runs of the project there should be enough time. I'll try to cleanup the code and check if it is possible to make it available for the public. But only if someone is interested.

BTW kryo and FST are not working. With these the speedup would be much bigger ...

**Comment by gdubya** [comment]
I'm interested! :+1:

**Comment by nbartels** [comment]
Can you tell something about the speed up?

**Comment by Gael** [comment]
Yes, the CompilationUnits required for test startup are stored in a file under folder gwt-Unitcache in the test classpath. So every time you start a test, gwt-test-utils relies on gwt internal to load the CompilationState from this cache if nothing as changed, and only rebuild what is needed in case of change.
Only the first time is slow since GWT needs to parse every jar in the classpath.

**Comment by nbartels** [comment]
Sounds really great. I give it a try and hope to speed up the unittest runs with this change even more. My solution is some how okay, but the (de)serialisation takes a long time. So maybe the gwt-Unitcache is more efficient.

**Comment by nbartels** [comment]
Okay, I did a short benchmark and really like the gwt-unitcache.
Without any caching the tests run about 29-30 secs each. With the new gwt-unitcache, the same tests run on the same machine with similar load only 8-9 secs.

So it's only 1/3 of the time used before. This is really awesome. Many thanks for this solution.
BTW I backported the change to the gwt-test-utils 0.52 and we use this modified version now. Maybe the times differ on newer gwt versions and newer gwt-test-utils.
